File: POO/TP/POO_TP2/src/ChristmasTree.java
public class ChristmasTree {

    // Dessine l'arbre.
    // `height` définit la hauteur de l'arbre entre la pointe et le tronc.
    public static void drawTree(int height, int titleBoxWidth, boolean padding) {
        // `midWidth` et `maxWidth` sont les valeurs de largeur de l'arbre.
        int midWidth = height + 2;
        int maxWidth = midWidth * 2 + 1;

        // `extraPadding` ajoute des espaces de padding horizontal
        // si l'on veut centrer l'arbre par rapport à une largeur spécifique.
        // Ici, la largeur spécifique est celle de la titleBox qui sera au-dessus.
        int extraPadding = maxWidth > titleBoxWidth+2 || !padding ? 0 : (titleBoxWidth+2-maxWidth) / 2;

        // Elargit la titleBox si l'arbre est plus large qu'elle (largeur default: 50).
        titleBoxWidth = Math.max(maxWidth, titleBoxWidth);

        // Dessine la titleBox liée à l'arbre.
        Utils.drawTitleBox(titleBoxWidth, "T A D A A A A  ♪", "", 1);


        // Dessine l'arbre lui-même en fonction de la hauteur donnée.
        // Hauteur = de la pointe au tronc exclus.
        for (int row = 0; row < height; row++) {
            System.out.print(" ".repeat(extraPadding + midWidth - row));
            System.out.print("*".repeat(row * 2 + 1));
            System.out.println(" ".repeat(midWidth - row));
        }
        // Dessine le tronc de l'arbre (fixe, c'est un arbre très stable).
        System.out.println(" ".repeat(extraPadding + midWidth) + "*" + " ".repeat(midWidth));
        System.out.println(" ".repeat(extraPadding + midWidth - 1) + "***" + " ".repeat(midWidth-1) + "\n");
    }
    public static void main(String[] args) {
        System.out.println();
        Utils.drawTitleBox(48, "L E  S A P I N  D E  N O Ë L", "POO | TP n°1", 1);
        int retry = -1;
        while (retry != 0) {

            // Demande la hauteur de l'arbre à l'utilisateur.
            int height = Utils.askForInt(4, 25, "Hauteur de l'arbre ? → ", false);
            System.out.println();
            // Dessine l'arbre avec la hauteur donnée.
            drawTree(height, 48, true);

            // Retry: 1 relance la fonction depuis le début (boucle while), 0 l'arrête
            // et revient au menu précédent.
            System.out.println("Voulez-vous dessiner un autre arbre ?\n");
            Menu.runMenu(true, null);
            r
            System.out.println();
        }
    }
}

File: POO/TP/POO_TP3/TP3/christmastree.py
from TP3 import utils, menu

# L'arbre est contenu dans une classe pour plus de simplicité et de lisibilité au niveau des variables.
class ChristmasTree:
    # `height` définit la hauteur de l'arbre, `spaceChar` définit le caractère utilisé hors de l'arbre,
    # `fillChar` définit le caractère utilisé hors de l'arbre et `basePadding` permet de décaler
    # l'arbre du bord de l'écran (s'applique à gauche et à droite).
    def __init__(self, height: int, spaceChar: str = " ", fillChar: str="*", basePadding:int=3):
        self.height = height
        self.spaceChar = spaceChar[0]
        self.fillChar = fillChar
        self.basePadding = basePadding

    # Si une textBox est renseignée et que l'arbre est plus petit que celle-ci, `centerWithBox` permet
    # d'ajouter un padding supplémentaire pour centrer l'arbre par rapport au centre de la textBox.
    # Si la box est plus grande que l'arbre, `textBoxIsResizable`=True permet d'agrandir la textBox
    # à la largeur max de l'arbre.
    def drawTree(self, textBox:menu.Box=None, centerWithBox=True, textBoxIsResizable=False):
        # `maxWidth` représente la largeur maximale de l'arbre [AVEC SON `basePadding`].
        maxWidth = self.height * 2 + self.basePadding*2 - 1
        extraPadding = 0

        if bool(textBox):
            # Centre l'arbre par rapport à la box.
            if centerWithBox is True and maxWidth < textBox.width+2:
                extraPadding = int(textBox.width+2-maxWidth)
            # Redimensionne la box par rapport à l'arbre.
            if textBoxIsResizable is True:
                textBox.width = max(textBox.width, maxWidth-2)

            # Dessine la textBox si elle a été renseignée.
            textBox.drawBox()
            utils.jumpLines()

        # Dessine l'arbre lui-même.
        for row in range(self.height):
            print((self.fillChar*(row*2+1)).center(maxWidth+extraPadding, self.spaceChar))
        print(self.fillChar.center(maxWidth+extraPadding))
        print((self.fillChar*3).center(maxWidth+extraPadding))
        utils.jumpLines()

def main():
    userRetry = -1
    while userRetry != 0:
        menu.Box("L E  S A P I N  D E  N O Ë L").drawBox()
        utils.jumpLines()

        # Demande à l'utilisateur la taille de l'arbre (Min: 4, max: 25).
        treeHeight = utils.getUserInput(4, 25, "Hauteur de l'arbre ? → ")
        utils.jumpLines()

        ChristmasTree(treeHeight).drawTree(menu.Box("T A D A A A A  ♪", title=""), textBoxIsResizable=True)

        # Menu retry (0 = quitter, 1 = réessayer).
        userRetry = menu.drawBinaryMenu("Voulez vous dessiner un autre arbre ?")
        utils.jumpLines()